The default number of threads is wrong within a distributed environment

### What happens?

When using duckdb from a Docker container in a Kubernetes cluster, the default number of threads used is the number of physical cores of the cluster, instead of the number of accessible cores. This results in oversubscribing issues, where there are far too many threads spawned compared to the number of cores to be used.

This is probably due to duckdb using shell commands like `nproc` to access the number of cores. Such commands are not cgroup aware and therefore yield incorrect results.

On the other hand, libraries like joblib inspect cgroup to give an accurate number of accessible cores. See [this section of joblib](https://github.com/joblib/loky/blob/master/loky/backend/context.py#L125).

cc @ogrisel that spotted the bug

### To Reproduce

Run duckdb in a Kubernetes pod and see the number of spawned threads with `htop`.
```python
import duckdb
con = duckdb.connect("duckdb.db")
# any sql query
```

Also run `nproc` in the shell (this will also give the number observed threads)

Then, install joblib:
```shell
pip install joblib
python -c "import joblib; print(joblib.cpu_count())"
```
This will give the number of accessible cores.
Alternatively, directly query cgroup attributes like joblib:
```shell
expr $(cat /sys/fs/cgroup/cpu/cpu.cfs_quota_us) / $(cat /sys/fs/cgroup/cpu/cpu.cfs_period_us)
```

### OS:

Ubuntu 20.04.4 LTS

### DuckDB Version:

0.7.0

### DuckDB Client:

Python 3.10

### Full Name:

Vincent Maladiere

### Affiliation:

Inria • scikit-learn

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

BTW, I get the following warning when building this branch:

```
In file included from /io/build/src/main/ub_duckdb_main.cpp:8:
/io/src/main/config.cpp: In function 'duckdb::idx_t duckdb::CGroupBandwidthQuota(duckdb::idx_t, duckdb::FileSystem&)':
/io/src/main/config.cpp:234:45: warning: format '%lld' expects argument of type 'long long int*', but argument 3 has type 'int64_t*' {aka 'long int*'} [-Wformat=]
  234 |                 if (sscanf(byte_buffer, "%lld %lld", &quota, &period) != 2) {
      |                                          ~~~^        ~~~~~~
      |                                             |        |
      |                                             |        int64_t* {aka long int*}
      |                                             long long int*
      |                                          %ld
/io/src/main/config.cpp:234:50: warning: format '%lld' expects argument of type 'long long int*', but argument 4 has type 'int64_t*' {aka 'long int*'} [-Wformat=]
  234 |                 if (sscanf(byte_buffer, "%lld %lld", &quota, &period) != 2) {
      |                                               ~~~^           ~~~~~~~
      |                                                  |           |
      |                                                  |           int64_t* {aka long int*}
      |                                                  long long int*
      |                                               %ld
/io/src/main/config.cpp:246:45: warning: format '%lld' expects argument of type 'long long int*', but argument 3 has type 'int64_t*' {aka 'long int*'} [-Wformat=]
  246 |                 if (sscanf(byte_buffer, "%lld", &quota) != 1) {
      |                                          ~~~^   ~~~~~~
      |                                             |   |
      |                                             |   int64_t* {aka long int*}
      |                                             long long int*
      |                                          %ld
/io/src/main/config.cpp:255:45: warning: format '%lld' expects argument of type 'long long int*', but argument 3 has type 'int64_t*' {aka 'long int*'} [-Wformat=]
  255 |                 if (sscanf(byte_buffer, "%lld", &period) != 1) {
      |                                          ~~~^   ~~~~~~~
      |                                             |   |
      |                                             |   int64_t* {aka long int*}
      |                                             long long int*
      |                                          %ld

```

--- a/src/main/config.cpp
+++ b/src/main/config.cpp
@@ -214,9 +214,71 @@ void DBConfig::SetDefaultMaxMemory() {
 	}
 }
 
+idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
+	static constexpr const char *cpu_max = "/sys/fs/cgroup/cpu.max";
+	static constexpr const char *cfs_quota = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us";
+	static constexpr const char *cfs_period = "/sys/fs/cgroup/cpu/cpu.cfs_period_us";
+
+	int64_t quota, period;
+	char byte_buffer[1000];
+	unique_ptr<FileHandle> handle;
+	int64_t read_bytes;
+
+	if (fs.FileExists(cpu_max)) {
+		// cgroup v2
+		// https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html
+		handle =
+		    fs.OpenFile(cpu_max, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK, FileSystem::DEFAULT_COMPRESSION);
+		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
+		byte_buffer[read_bytes] = '\0';
+		if (sscanf(byte_buffer, "%lld %lld", &quota, &period) != 2) {
+			return physical_cores;
+		}
+	} else if (fs.FileExists(cfs_quota) && fs.FileExists(cfs_period)) {
+		// cgroup v1
+		// https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html#management
+
+		// Read the quota, this indicates how many microseconds the CPU can be utilized by this cgroup per period
+		handle = fs.OpenFile(cfs_quota, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK,
+		                     FileSystem::DEFAULT_COMPRESSION);
+		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
+		byte_buffer[read_bytes] = '\0';
+		if (sscanf(byte_buffer, "%lld", &quota) != 1) {
+			return physical_cores;
+		}
+
+		// Read the time period, a cgroup can utilize the CPU up to quota microseconds every period
+		handle = fs.OpenFile(cfs_period, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK,
+		                     FileSystem::DEFAULT_COMPRESSION);
+		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
+		byte_buffer[read_bytes] = '\0';
+		if (sscanf(byte_buffer, "%lld", &period) != 1) {
+			return physical_cores;
+		}
+	} else {
+		// No cgroup quota
+		return physical_cores;
+	}
+	if (quota > 0 && period > 0) {
+		return std::ceil(quota / period);
+	} else {
+		return physical_cores;
+	}
+}
+
+idx_t GetSystemMaxThreadsInternal(FileSystem &fs) {
+	idx_t physical_cores = std::thread::hardware_concurrency();
+#ifdef __linux__
+	auto cores_available_per_period = CGroupBandwidthQuota(physical_cores, fs);
+	return physical_cores;
+#else
+	return physical_cores;
+#endif
+}
+
 void DBConfig::SetDefaultMaxThreads() {
 #ifndef DUCKDB_NO_THREADS
-	options.maximum_threads = std::thread::hardware_concurrency();
+	options.maximum_threads = GetSystemMaxThreadsInternal(*file_system);
 #else
 	options.maximum_threads = 1;
 #endif

--- a/src/main/config.cpp
+++ b/src/main/config.cpp
@@ -9,6 +9,9 @@
 #include "duckdb/common/thread.hpp"
 #endif
 
+#include <cstdio>
+#include <inttypes.h>
+
 namespace duckdb {
 
 #define DUCKDB_GLOBAL(_PARAM)                                                                                          \
@@ -214,7 +217,6 @@ void DBConfig::SetDefaultMaxMemory() {
 	}
 }
 
-#ifdef __linux__
 idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 	static constexpr const char *cpu_max = "/sys/fs/cgroup/cpu.max";
 	static constexpr const char *cfs_quota = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us";
@@ -232,7 +234,7 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		    fs.OpenFile(cpu_max, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK, FileSystem::DEFAULT_COMPRESSION);
 		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
 		byte_buffer[read_bytes] = '\0';
-		if (sscanf(byte_buffer, "%ld %ld", &quota, &period) != 2) {
+		if (std::sscanf(byte_buffer, "%" SCNd64 " %" SCNd64 "", &quota, &period) != 2) {
 			return physical_cores;
 		}
 	} else if (fs.FileExists(cfs_quota) && fs.FileExists(cfs_period)) {
@@ -244,7 +246,7 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		                     FileSystem::DEFAULT_COMPRESSION);
 		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
 		byte_buffer[read_bytes] = '\0';
-		if (sscanf(byte_buffer, "%ld", &quota) != 1) {
+		if (std::sscanf(byte_buffer, "%" SCNd64 "", &quota) != 1) {
 			return physical_cores;
 		}
 
@@ -253,7 +255,7 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		                     FileSystem::DEFAULT_COMPRESSION);
 		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
 		byte_buffer[read_bytes] = '\0';
-		if (sscanf(byte_buffer, "%ld", &period) != 1) {
+		if (std::sscanf(byte_buffer, "%" SCNd64 "", &period) != 1) {
 			return physical_cores;
 		}
 	} else {
@@ -266,7 +268,6 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		return physical_cores;
 	}
 }
-#endif
 
 idx_t GetSystemMaxThreadsInternal(FileSystem &fs) {
 	idx_t physical_cores = std::thread::hardware_concurrency();

--- a/src/main/config.cpp
+++ b/src/main/config.cpp
@@ -270,7 +270,7 @@ idx_t GetSystemMaxThreadsInternal(FileSystem &fs) {
 	idx_t physical_cores = std::thread::hardware_concurrency();
 #ifdef __linux__
 	auto cores_available_per_period = CGroupBandwidthQuota(physical_cores, fs);
-	return physical_cores;
+	return MaxValue<idx_t>(cores_available_per_period, 1);
 #else
 	return physical_cores;
 #endif

--- a/src/main/config.cpp
+++ b/src/main/config.cpp
@@ -214,6 +214,7 @@ void DBConfig::SetDefaultMaxMemory() {
 	}
 }
 
+#ifdef __linux__
 idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 	static constexpr const char *cpu_max = "/sys/fs/cgroup/cpu.max";
 	static constexpr const char *cfs_quota = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us";
@@ -231,7 +232,7 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		    fs.OpenFile(cpu_max, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK, FileSystem::DEFAULT_COMPRESSION);
 		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
 		byte_buffer[read_bytes] = '\0';
-		if (sscanf(byte_buffer, "%lld %lld", &quota, &period) != 2) {
+		if (sscanf(byte_buffer, "%ld %ld", &quota, &period) != 2) {
 			return physical_cores;
 		}
 	} else if (fs.FileExists(cfs_quota) && fs.FileExists(cfs_period)) {
@@ -243,7 +244,7 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		                     FileSystem::DEFAULT_COMPRESSION);
 		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
 		byte_buffer[read_bytes] = '\0';
-		if (sscanf(byte_buffer, "%lld", &quota) != 1) {
+		if (sscanf(byte_buffer, "%ld", &quota) != 1) {
 			return physical_cores;
 		}
 
@@ -252,7 +253,7 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		                     FileSystem::DEFAULT_COMPRESSION);
 		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
 		byte_buffer[read_bytes] = '\0';
-		if (sscanf(byte_buffer, "%lld", &period) != 1) {
+		if (sscanf(byte_buffer, "%ld", &period) != 1) {
 			return physical_cores;
 		}
 	} else {
@@ -265,6 +266,7 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		return physical_cores;
 	}
 }
+#endif
 
 idx_t GetSystemMaxThreadsInternal(FileSystem &fs) {
 	idx_t physical_cores = std::thread::hardware_concurrency();

--- a/src/main/config.cpp
+++ b/src/main/config.cpp
@@ -261,7 +261,7 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		return physical_cores;
 	}
 	if (quota > 0 && period > 0) {
-		return std::ceil(quota / period);
+		return idx_t(std::ceil((double)quota / (double)period));
 	} else {
 		return physical_cores;
 	}

--- a/src/main/config.cpp
+++ b/src/main/config.cpp
@@ -218,31 +218,31 @@ void DBConfig::SetDefaultMaxMemory() {
 }
 
 idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
-	static constexpr const char *cpu_max = "/sys/fs/cgroup/cpu.max";
-	static constexpr const char *cfs_quota = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us";
-	static constexpr const char *cfs_period = "/sys/fs/cgroup/cpu/cpu.cfs_period_us";
+	static constexpr const char *CPU_MAX = "/sys/fs/cgroup/cpu.max";
+	static constexpr const char *CFS_QUOTA = "/sys/fs/cgroup/cpu/cpu.cfs_quota_us";
+	static constexpr const char *CFS_PERIOD = "/sys/fs/cgroup/cpu/cpu.cfs_period_us";
 
 	int64_t quota, period;
 	char byte_buffer[1000];
 	unique_ptr<FileHandle> handle;
 	int64_t read_bytes;
 
-	if (fs.FileExists(cpu_max)) {
+	if (fs.FileExists(CPU_MAX)) {
 		// cgroup v2
 		// https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html
 		handle =
-		    fs.OpenFile(cpu_max, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK, FileSystem::DEFAULT_COMPRESSION);
+		    fs.OpenFile(CPU_MAX, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK, FileSystem::DEFAULT_COMPRESSION);
 		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
 		byte_buffer[read_bytes] = '\0';
 		if (std::sscanf(byte_buffer, "%" SCNd64 " %" SCNd64 "", &quota, &period) != 2) {
 			return physical_cores;
 		}
-	} else if (fs.FileExists(cfs_quota) && fs.FileExists(cfs_period)) {
+	} else if (fs.FileExists(CFS_QUOTA) && fs.FileExists(CFS_PERIOD)) {
 		// cgroup v1
 		// https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html#management
 
 		// Read the quota, this indicates how many microseconds the CPU can be utilized by this cgroup per period
-		handle = fs.OpenFile(cfs_quota, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK,
+		handle = fs.OpenFile(CFS_QUOTA, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK,
 		                     FileSystem::DEFAULT_COMPRESSION);
 		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
 		byte_buffer[read_bytes] = '\0';
@@ -251,7 +251,7 @@ idx_t CGroupBandwidthQuota(idx_t physical_cores, FileSystem &fs) {
 		}
 
 		// Read the time period, a cgroup can utilize the CPU up to quota microseconds every period
-		handle = fs.OpenFile(cfs_period, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK,
+		handle = fs.OpenFile(CFS_PERIOD, FileFlags::FILE_FLAGS_READ, FileSystem::DEFAULT_LOCK,
 		                     FileSystem::DEFAULT_COMPRESSION);
 		read_bytes = fs.Read(*handle, (void *)byte_buffer, 999);
 		byte_buffer[read_bytes] = '\0';

